src/toxcore/tcp/server/server.rs

```diff
@@ -54,14 +54,24 @@ impl Server {
     pub fn set_udp_onion_sink(&mut self, onion_sink: mpsc::Sender<(OnionRequest, SocketAddr)>) {
         self.onion_sink = Some(onion_sink)
     }
-    /** Insert the client into connected_clients. Do nothing else.
+    /** Insert the client into `connected_clients`. If `connected_clients`
+    contains a client with the same pk it will be terminated.
     */
-    pub fn insert(&self, client: Client) {
+    pub fn insert(&self, client: Client) -> IoFuture<()> {
         let mut state = self.state.write();
+
+        let shutdown_future = if state.connected_clients.contains_key(&client.pk()) {
+            self.shutdown_client_inner(&client.pk(), &mut state)
+        } else {
+            Box::new(future::ok(()))
+        };
+
         state.keys_by_addr
             .insert((client.ip_addr(), client.port()), client.pk());
         state.connected_clients
             .insert(client.pk(), client);
```

Comment:
> state.connected_clients.insert

And you will replace the existing client without disconnecting it.

Reply:
Nope, I called `shutdown_client_inner` for it. So there won't be a client with the same PK anymore.

Reply:
You are right